### PR TITLE
PEP 594: add myself as co-author and target for Python 3.11

### PR DESCRIPTION
**Pull request by brettcannon**

### CODE REVIEW

File: pep-0594.rst
Comment (JelleZijlstra) [inline pep-0594.rst:96]:
```suggestion
  kids and can make crazy sounds. Making a computer generate sounds is a
```

File: pep-0594.rst
Comment (JelleZijlstra) [inline pep-0594.rst:52]:
```suggestion
Times have changed. The introduction of the PyPI (née Cheeseshop), setuptools,
```
Feel free to ignore, English allows dropping the accent

File: pep-0594.rst
Comment (JelleZijlstra) [inline pep-0594.rst:114]: Any reason to remove the verbiage about a feature freeze?

**Review by JelleZijlstra — COMMENTED** [review]
Thanks, a couple of pieces of feedback

**Review by JelleZijlstra — APPROVED** [review]
Thanks for helping to push this forward!

**Comment by gvanrossum** [comment]
Does it sound like the SC would accept this? That would be great!

File: pep-0594.rst
Comment (JelleZijlstra) [inline pep-0594.rst:930]: The PEP lists lib2to3 among modules to be kept, but it's been independently scheduled for removal in 3.13 already: https://docs.python.org/3.10/library/2to3.html#module-lib2to3.

Also (speaking as Black maintainer), it's not exactly true that Black depends on lib2to3; instead we use a fork of the module that has been enhanced to support match/case. So from Black's perspective, there's no need to keep it in the stdlib.

**Comment by CAM-Gerlach** [comment]
Excited to see this finally moving forward again! Happy to assist in seeing this through; if I can help with PEP updates/editing or the more mechanical tasks in the CPython repo, feel free to ping me here or on the Discourse thread.

Review in progress...

File: pep-0594.rst
Comment (CAM-Gerlach) [inline pep-0594.rst:137]:
```suggestion
    asynchat,**3.6** (3.0\*),3.12,1999,**yes**,asyncio_
    asyncore,**3.6** (3.0\*),3.12,1999,**yes**,asyncio_
```

These modules (and `smtpd`) had formal `DeprecationWarnings` added in 3.10 per [bpo-44498](https://bugs.python.org/issue44498), were attempted to be removed in 3.11 but the change was reverted after you refereed it to the SC in python/steering-council#86 and a PEP 387 exception was rejected, so per policy this will now land in 3.12. See [bpo-28533](https://bugs.python.org/issue28533) for others curious.

File: pep-0594.rst
Comment (CAM-Gerlach) [inline pep-0594.rst:154]:
```suggestion
    smtpd,"**3.4.7**, **3.5.4**",3.12,2001,**yes**,"aiosmtpd_"
```

Same as the `asyn*` modules, `smtpd` had formal `DeprecationWarnings` added in 3.10 per [bpo-4498](https://bugs.python.org/issue44498), were attempted to be removed in 3.11 in [bpo-28533](https://bugs.python.org/issue28533) but the change was reverted and rescheduled—presumably for 3.12, right?

File: pep-0594.rst
Comment (CAM-Gerlach) [inline pep-0594.rst:533]: I suggest removing `optparse` as a replacement for `getopt`; you could consider adding Click to both as another replacement (which uses a private fork of `optparse` under the hook to provide its features/behavior). You could also add [pyupgrade](https://github.com/asottile/pyupgrade) as a possible replacement for `lib2to3`. I also suggest removing `colour`, as it has [no commits since 2017](https://github.com/vaab/colour) and relies on a bunch of deprecated/EoL packaging stuff (EOL d21, horrific setup.py hacks, etc).

File: pep-0594.rst
Comment (CAM-Gerlach) [inline pep-0594.rst:153]:
```suggestion
    pipes,3.11,3.13,1992,no,"subprocess_"
```

I'm guessing this was a typo, since it was 3.8 before like the others and I don't see a deprecated noted in the [docs](https://docs.python.org/3.11/library/pipes.html)

File: pep-0594.rst
Comment (CAM-Gerlach) [inline pep-0594.rst:483]: I suggest moving this to the `Removed` section, since this was removed on schedule in Python 3.10.

File: pep-0594.rst
Comment (CAM-Gerlach) [inline pep-0594.rst:519]: You could either consider deleting this section, and `formatter` + `fpectl` entirely since they were already removed from the released version of Python, or move them the latter two here—either way, best to be consistent.

File: pep-0594.rst
Comment (CAM-Gerlach) [inline pep-0594.rst:144]:
```suggestion
    formatter,**3.4**,**3.10**,1995,no,\-
```

Was already removed on schedule in Python 3.10, so bolding accordingly

File: pep-0594.rst
Comment (CAM-Gerlach) [inline pep-0594.rst:156]:
```suggestion
    spwd,3.11,3.13,2005,no,"python-pam_"
```

Simplepam only had [one PyPI release in 2014](https://pypi.org/project/simplepam/#history) and [no commits since then(https://github.com/leonnnn/python3-simplepam).

File: pep-0594.rst
Comment (CAM-Gerlach) [inline pep-0594.rst:162]: You'll want to eliminate any alternatives cut from the list above here too

File: pep-0594.rst
Comment (CAM-Gerlach) [inline pep-0594.rst:143]:
```suggestion
    crypt,3.11,3.13,1994,**yes (inactive)**,"legacycrypt_, bcrypt_, argon2-cffi_, hashlib_, passlib_"
```

use the canonical PyPI package name for `argon2-cffi`

File: pep-0594.rst
Comment (CAM-Gerlach) [inline pep-0594.rst:163]:
```suggestion
.. _argon2-cffi: https://pypi.org/project/argon2-cffi/
```

Use the canonical PyPI package name, to match the above.

**Review by CAM-Gerlach — CHANGES_REQUESTED** [review]
Thanks again, @brettcannon ! I've got a mix of small textual suggestions and substantive content questions/comments. The one thing I didn't check was maintainer/expert status of the modules, as that would be rather time-intensive, but I can if that would be helpful. Thanks!

File: pep-0594.rst
Comment (CAM-Gerlach) [reply pep-0594.rst:930]: Also mentioned [here](https://github.com/python/peps/pull/2262#discussion_r791303079), as noted in my review

**Comment by CAM-Gerlach** [comment]
Also, @brettcannon , don't forget to add yourself to `CODEOWNERS` :)

File: pep-0594.rst
Comment (CAM-Gerlach) [inline pep-0594.rst:561]: The above line (one higher than I can reach with GitHub Suggestions) should also be updated to remove ``colour``, assuming it is removed from the table as suggested

File: pep-0594.rst
Comment (CAM-Gerlach) [reply pep-0594.rst:605]: Also, for what its worth now (probably not much), it is [listed in PEP 360](https://www.python.org/dev/peps/pep-0360/#optik) as being externally maintained upstream, which has been dead for over 15 years and a practice which that PEP strongly discourages for existing modules and prohibits in the future, the four remaining examples (Expat, `etree`, `optparse` and `wsgiref`) being grandfathered in by necessity at the time.

File: pep-0594.rst
Comment (hugovk) [reply pep-0594.rst:108]: See the earlier, now deleted text, it talked of 3.9 EOL:

> All other deprecated modules are fully supported and will receive security
updates until Python 3.9 reaches its end-of-life. Python 3.9.0 will
be released about 18 months after 3.8.0 (April 2021?) and most likely
be supported for five years after the release. The estimated EOL of Python 3.9
is in 2026. 

Lines 126-130.

File: pep-0594.rst
Comment (CAM-Gerlach) [inline pep-0594.rst:532]: Just for reference: grep.app hits for ``\n(import|from) {module}` for each of the modules:

* `colorsys`:
  993
* `fileinput`:
  1478
* `getopt`:
  4862
* `optparse`:
  12339
* `wave`:
  1569

And for their replacements:

* `argparse`:
  128400
* `colour`:
  769
* `colormath`:
  53
* `colorspacious`:
  7
* `pillow` (`import PIL`):
  39500

As an interesting twist, a couple of conversion functions in `colorsys` are actually used [one place](https://github.com/python-pillow/Pillow/blob/main/src/PIL/ImageColor.py) in Pillow (though they could easily be copied in, of course).

File: pep-0594.rst
Comment (CAM-Gerlach) [reply pep-0594.rst:108]: Right, so the equivalent EOL should be that of Python 3.12, not 3.11, as mentioned above.

Comment (brettcannon) [reply pep-0594.rst:108]: This is accurate and on purpose. The deprecations would start in Python 3.11 and thus how long people have to notice that the deprecations have occurred. Otherwise it is not an important detail and it can be removed.

File: pep-0594.rst
Comment (brettcannon) [reply pep-0594.rst:113]: The version that the module will be removed in is Python 3.13, not 3.12. And you can just consider it an FYI on when 3.11 will go away, but it's also the maximum length of time one can go and avoid the warning on a supported version of Python.

File: pep-0594.rst
Comment (brettcannon) [reply pep-0594.rst:207]: I missed it. There was a lot to tweak and only so much time 😉.

File: pep-0594.rst
Comment (brettcannon) [reply pep-0594.rst:533]: The key point of the suggestions isn't necessarily the _best_ option, but to prove the point that there _is_ a serviceable option. That means optparse is a legitimate replacement for getopt.

File: pep-0594.rst
Comment (brettcannon) [reply pep-0594.rst:903]: I'm not concerned. They are not a key part of the PEP and finding the links later is not difficult.

File: pep-0594.rst
Comment (brettcannon) [reply pep-0594.rst:605]: The decisions made by this PEP should not be discussed here as that is not the point of this PR. Once this PR gets merged and I review the comments on discuss.python.org around `cgi` then I will post the latest draft for another round of discussion where you may bring up removing optparse (i.e. I don't want to spark that argument 😉; you weren't around for the first round, but it got heated and was extremely draining).

**Review by JelleZijlstra — APPROVED** [review]

**Comment by brettcannon** [comment]
@JelleZijlstra I unfortunately wasn't done with this PR (I'm actually actively editing it right now and pushing my branch as I go), hence the unresolved conversations.

I'll create a separate PR for the changes I'm currently making.

**Comment by JelleZijlstra** [comment]
Sorry for that! I interpreted the review request as an indication that you were ready.

**Comment by brettcannon** [comment]
I actually didn't make the review request; GitHub did automatically when I edited `CODEOWNERS`.

File: pep-0594.rst
Comment (CAM-Gerlach) [reply pep-0594.rst:108]: I'm a little confused, sorry. How does the Python 3.11 _EOL_ affect how long people have to notice that deprecations have occurred? The Python 3.11 EOL doesn't change anything, right, since the situation is exactly the same in Python 3.12 (as you note in the PEP)?

**Comment by CAM-Gerlach** [comment]
Just FYI @brettcannon @JelleZijlstra marking the PR as Draft is very useful for situations like these, to prevent others merging it while you're working on it and send a clear signal that it is a work in progress.

File: pep-0594.rst
Comment (CAM-Gerlach) [reply pep-0594.rst:113]: Thanks for explaining—but unfortunately, I'm still a little confused. Python 3.11 will _have_ the warnings, right? The last supported version that does _not_ have the warnings is 3.10, so did you mean that EOL instead?

File: pep-0594.rst
Comment (JelleZijlstra) [reply pep-0594.rst:108]: I'm with CAM, I feel like the relevant dates are the EOL of 3.10 (last version without warnings) and the EOL of 3.12 (last version where the modules still exist).

File: pep-0594.rst
Comment (CAM-Gerlach) [reply pep-0594.rst:207]: Thanks :slightly_smiling_face: Let me know if I can help

File: pep-0594.rst
Comment (CAM-Gerlach) [reply pep-0594.rst:605]: Sorry, my fault. It started out as just a brief mention as something to consider bringing up in the discussion, and morphed into a much longer note, mostly to myself, with some numbers and rationale for reference. I should have moved it to my personal scratchpad, it just got late and I was hitting the green button on everything and didn't think it through.

For what its worth, I figured its probably not worth getting this PEP waylaid over given the likely backlash, but wanted to give it due consideration.

File: pep-0594.rst
Comment (CAM-Gerlach) [reply pep-0594.rst:533]: Understood, thanks.

File: pep-0594.rst
Comment (CAM-Gerlach) [reply pep-0594.rst:627]: Resolved in #2274

File: pep-0594.rst
Comment (CAM-Gerlach) [reply pep-0594.rst:483]: Resolved in #2274

File: pep-0594.rst
Comment (CAM-Gerlach) [reply pep-0594.rst:519]: Resolved in #2274

File: pep-0594.rst
Comment (CAM-Gerlach) [reply pep-0594.rst:108]: Since it seems we have consensus here, I've followed this up with PR #2290

File: pep-0594.rst
Comment (CAM-Gerlach) [reply pep-0594.rst:113]: See PR #2290 to fix this

**Comment by brettcannon** [comment]
@CAM-Gerlach correct, but the PR was submitted in a form I was ready to merge. You just happened to disagree. 😉 So this wasn't an issue about me putting up a PR I wasn't to have merged from the start, but Jelle seeing a buggy signal from GitHub that somehow I was ready for another round of reviews and thus was done. I don't think constantly flipping a PR from draft to ready while addressing reviews is a common practice.

**Comment by CAM-Gerlach** [comment]
Oh sorry, I wasn't referring to my review comments at all—what I meant was what what I presumed were other additional changes you were continuing to iterate on yourself (the `cgi` stuff?) that I thought you were referring to in your reply to Jelle mentioning that the merge was premature—

> @JelleZijlstra I unfortunately wasn't done with this PR (I'm actually actively editing it right now and pushing my branch as I go)

But maybe I misunderstood what you were saying there? I certainly wasn't suggesting flipping draft mode back and forth when responding to discussions, just while you had changes _you_ wanted to make on the PR (or there was some other important reason why it shouldn't be merged once approved).

For avoiding merges when there are unresolved discussions, we could consider enabled the corresponding feature of GitHub branch protection which puts an alert and a count if there are unresolved review comments in the merge box, and ensures discussions are resolved (or the check is explicitly overriden) before merging. It really has come in handy on our various repos, since it can get hard to keep track (even by those involved) if there are a lot of discussions or a lag between them.

**Comment by CAM-Gerlach** [comment]
Oh okay, gotcha. Sorry I misunderstood and thanks for clarifying!

**Pull request merged**
